tla/tlc.py

```diff
@@ -0,0 +1,218 @@
+    )
+    mc.add_argument(
+        "--raft-configs",
+        type=str,
+        default="1C2N",
+        help="Raft configuration sequence, defaults to 1C2N",
+    )
+    mc.add_argument(
+        "--disable-check-quorum", action="store_true", help="Disable CheckQuorum action"
+    )
+
+    # Trace validation
+    tv = subparsers.add_parser("tv", help="Trace validation")
+    # DFS is a good default for trace validation
+    tv.add_argument(
+        "--disable-dfs",
+        action="store_true",
+        help="Set TLC to use depth-first search",
+    )
+    tv.add_argument(
+        "--driver-trace",
+        type=pathlib.Path,
+        default=None,
+        help="Path to a CCF Raft driver trace .ndjson file, produced by make_traces.sh",
+    )
+
+    # Simulation
+    sim = subparsers.add_parser("sim", help="Simulation")
+    sim.add_argument(
+        "--num",
+        type=int,
+        default=None,
+        help="Number of behaviours to simulate per worker thread",
+    )
+    sim.add_argument(
+        "--depth",
+        type=int,
+        default=500,
+        help="Set the depth of the simulation, defaults to 500",
+    )
+    sim.add_argument(
+        "--max-seconds",
+        type=int,
+        default=1200,
+        help="Set the timeout of the simulation, defaults to 1200 seconds",
+    )
+
+    parser.add_argument(
+        "spec", type=pathlib.Path, help="Path to the TLA+ specification"
+    )
+
+    return parser
+
+
+if __name__ == "__main__":
+    env = os.environ.copy()
+    args = cli().parse_args()
+    jvm_args = DEFAULT_JVM_ARGS
+    cp_args = DEFAULT_CLASSPATH_ARGS
+    tlc_args = [
+        "-workers",
+        args.workers,
+        "-checkpoint",
+        str(args.checkpoint),
+        "-lncheck",
+        args.lncheck,
+    ]
+    trace_name = args.trace_name or os.path.basename(args.config or args.spec).replace(
+        ".tla", ""
+    )
+
+    if "CI" in env:
+        # When run in CI, format output for GitHub, and participate in statistics collection
+        jvm_args.append("-Dtlc2.TLC.ide=Github")
+        jvm_args.append(
+            "-Dutil.ExecutionStatisticsCollector.id=be29f6283abeed2fb1fd0be898bc6601"
+        )
+
+    if args.trace_name or "CI" in env:
+        # When run in CI, or told explicitly, dump a trace
+        tlc_args.extend(["-dumpTrace", "json", f"{trace_name}.trace.json"])
+    if args.jmx:
+        jvm_args.append("-Dcom.sun.management.jmxremote")
+        jvm_args.append("-Dcom.sun.management.jmxremote.port=55449")
+        jvm_args.append("-Dcom.sun.management.jmxremote.ssl=false")
+        jvm_args.append("-Dcom.sun.management.jmxremote.authenticate=false")
+    if not args.disable_cdot:
+        jvm_args.append("-Dtlc2.tool.impl.Tool.cdot=true")
+    if args.config is not None:
+        tlc_args.extend(["-config", args.config])
+    if args.dot:
+        tlc_args.extend(
+            ["-dump", "dot,constrained,colorize,actionlabels", f"{trace_name}.dot"]
```

Comment:
can you remind me that constrained does? cannot see it in https://github.com/tlaplus/tlaplus/blob/master/general/docs/current-tools.md#command-line-options

Reply:
https://github.com/tlaplus/tlaplus/commit/24479f74c8a02fbe2f83981ca6210e2f7fa7799f

Reply:
thanks for clarifying

Reply:
Upstream: https://github.com/tlaplus/tlaplus/pull/1022